### PR TITLE
REFACTOR-#5951: remove code duplication for `to_pickle_distributed`

### LINKED ISSUE #5951: CLN: remove code duplication for `to_pickle_distributed`

### PR DESCRIPTION
**Pull request by anmyachev**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #5951 
- [ ] tests added and passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/io/pickle/pickle_dispatcher.py [hunk 1]
```diff
@@ -89,3 +92,50 @@ def _read(cls, filepath_or_buffer, **kwargs):
         return cls.query_compiler_cls(
             cls.frame_cls(partition_ids, new_index, new_columns)
         )
+
+    @classmethod
+    def _to_method(cls, qc, **kwargs):
```
Comment (dchigarev) [inline modin/core/io/pickle/pickle_dispatcher.py:97]: why does the name of the method so strange?
```suggestion
    def _to_pickle(cls, qc, **kwargs):
```

**Review by dchigarev — APPROVED** [review]
LGTM!

Left one comment regarding naming

File: modin/core/io/pickle/pickle_dispatcher.py (hunk 1 above)
Comment (anmyachev) [reply modin/core/io/pickle/pickle_dispatcher.py:97]: My guess that it should be a generic name like "_read" for any read function so I definitely would not want to add the format name. Maybe just `_to`?

Comment (dchigarev) [reply modin/core/io/pickle/pickle_dispatcher.py:97]: maybe something like `export()`?

by the way, why does the method's name starts with `_`, the `.read` name doesn't do so

Comment (anmyachev) [reply modin/core/io/pickle/pickle_dispatcher.py:97]: `write` looks good.

`read` is a base class method (`FileDispatcher`) that calls the `_read` method of the child class. As long as there is no such structure for write functions, methods can be called without `_`, ok.

**Comment by anmyachev** [comment]
@dchigarev thanks for review! It's ready for another one.

**Review by dchigarev — DISMISSED** [review]

File: modin/core/io/pickle/pickle_dispatcher.py [hunk 2]
```diff
@@ -89,3 +92,50 @@
         return cls.query_compiler_cls(
             cls.frame_cls(partition_ids, new_index, new_columns)
         )
+
+    @classmethod
+    def write(cls, qc, **kwargs):
```
[the remark this answers is not in the document]
Comment (YarShev) [reply modin/core/io/pickle/pickle_dispatcher.py:97]: Should we fix this alert?

File: modin/core/io/pickle/pickle_dispatcher.py [hunk 3]
```diff
@@ -89,3 +92,50 @@ def _read(cls, filepath_or_buffer, **kwargs):
...
+            to run ``to_pickle_distributed`` on.
+        **kwargs : dict
+            Parameters for ``pandas.to_pickle(**kwargs)``.
+        """
+        if not (
+            isinstance(kwargs["filepath_or_buffer"], str)
+            and "*" in kwargs["filepath_or_buffer"]
+        ) or not isinstance(qc, PandasQueryCompiler):
+            warnings.warn("Defaulting to Modin core implementation")
+            return cls.base_io.to_pickle(qc, **kwargs)
```
Comment (anmyachev) [inline modin/core/io/pickle/pickle_dispatcher.py:121]:
```suggestion
            cls.base_io.to_pickle(qc, **kwargs)
            return
```

File: modin/core/io/pickle/pickle_dispatcher.py [hunk 4]
```diff
@@ -89,3 +92,50 @@ def _read(cls, filepath_or_buffer, **kwargs):
...
+            The query compiler of the Modin dataframe that we want
+            to run ``to_pickle_distributed`` on.
+        **kwargs : dict
+            Parameters for ``pandas.to_pickle(**kwargs)``.
+        """
+        if not (
+            isinstance(kwargs["filepath_or_buffer"], str)
+            and "*" in kwargs["filepath_or_buffer"]
+        ) or not isinstance(qc, PandasQueryCompiler):
+            warnings.warn("Defaulting to Modin core implementation")
```
[the remark this answers is not in the document]
Comment (anmyachev) [reply modin/core/io/pickle/pickle_dispatcher.py:120]: This is to inform the user that the experimental functionality has not been used. In addition, core functionality may stop defaulting to pandas at some point.

[the remark this answers is not in the document]
Comment (YarShev) [reply modin/core/io/pickle/pickle_dispatcher.py:120]: We call `cls.base_io.to_pickle(qc, **kwargs)` that is defaulting to pandas anyway at the moment. I think the new added warning is redundant. Also, we should probably move ExperimentalPickleDispatcher into experimental IO.

File: modin/core/io/pickle/pickle_dispatcher.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (anmyachev) [reply modin/core/io/pickle/pickle_dispatcher.py:97]: Fixed

**Comment by anmyachev** [comment]
@YarShev thanks for review! It's ready for another one.

File: modin/core/io/pickle/pickle_dispatcher.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (anmyachev) [reply modin/core/io/pickle/pickle_dispatcher.py:120]: This warning was before, I still want to leave it like that.

> Also, we should probably move ExperimentalPickleDispatcher into experimental IO.

I can do it.

[the remark this answers is not in the document]
Comment (anmyachev) [reply modin/core/io/pickle/pickle_dispatcher.py:120]: Moved

**Comment by anmyachev** [comment]
@YarShev could we merge it?

**Review by YarShev — APPROVED** [review]

**Pull request merged**